builtin/logical/pki/storage.go

```diff
@@ -412,19 +415,22 @@ func resolveKeyReference(ctx context.Context, s logical.Storage, reference strin
 		return config.DefaultKeyId, nil
 	}
 
-	keys, err := listKeys(ctx, s)
-	if err != nil {
-		return keyID("list-error"), err
-	}
-
-	// Cheaper to list keys and check if an id is a match...
-	for _, keyId := range keys {
-		if keyId == keyID(reference) {
-			return keyId, nil
+	// Lookup by a direct get first to see if our reference is an ID, this is quick and cached.
+	if len(reference) == uuidLength {
+		entry, err := s.Get(ctx, keyPrefix+reference)
+		if err != nil {
+			return keyID("key-read"), err
+		}
+		if entry != nil {
+			return keyID(reference), nil
 		}
 	}
 
 	// ... than to pull all keys from storage.
+	keys, err := listKeys(ctx, s)
```

Comment:
Seeing this part in particular makes me think its not actually that great of an idea @sgmiller. We maybe save the `id` case slightly, but for a chunk of these calls, we'll likely have names which still incurs the list. We're just moving around the list and perhaps most of the time deferring it until after a failed `get`. 

Are failed `Get`'s cached?

I'm thinking a proper issuer caching mechanism would be better than this fix, at least in my opinion.